compositor_render/src/renderer/scene.rs

```diff
@@ -207,40 +216,94 @@ impl Scene {
-                    .collect::<Result<_, _>>()?;
-                let node =
-                    Node::new(ctx, transform, inputs).map_err(SceneUpdateError::RenderNodeError)?;
-                let node = Arc::new(node);
-                new_nodes.insert(node_id.clone(), node.clone());
-                return Ok(node);
+            let node_spec = spec.nodes.iter().find(|node| &node.node_id == node_id);
+            if let Some(node_spec) = node_spec {
+                for child_id in &node_spec.input_pads {
+                    Self::ensure_node(ctx, child_id, spec, inputs, new_nodes)?;
+                }
+                let node = Node::new(ctx, node_spec).map_err(SceneUpdateError::RenderNodeError)?;
+                new_nodes.insert(node_id.clone(), node);
+                return Ok(());
             }
         }
 
         // If there is no node with id node_id, assume it's an input. Pipeline validation should
         // make sure that scene does not refer to missing entities.
         let node = Node::new_input(node_id).map_err(SceneUpdateError::InputNodeError)?;
-        let node = Arc::new(node);
-        new_nodes.insert(node_id.clone(), node.clone());
-        inputs.insert(
-            InputId(node_id.clone()),
-            (node.clone(), InputTexture::new()),
-        );
-        Ok(node)
+        new_nodes.insert(node_id.clone(), node);
+        inputs.insert(InputId(node_id.clone()), InputTexture::new());
+        Ok(())
     }
 }
+
+pub struct SceneNodesSet {
+    nodes: HashMap<NodeId, Node>,
+}
+
+impl SceneNodesSet {
+    pub fn new() -> Self {
+        Self {
+            nodes: HashMap::new(),
+        }
+    }
+
+    pub fn node(&self, node_id: &NodeId) -> Result<&Node, SceneError> {
+        self.nodes
+            .get(node_id)
+            .ok_or_else(|| SceneError::MissingNode(node_id.clone()))
+    }
+
+    pub fn node_mut(&mut self, node_id: &NodeId) -> Result<&mut Node, SceneError> {
+        self.nodes
+            .get_mut(node_id)
+            .ok_or_else(|| SceneError::MissingNode(node_id.clone()))
+    }
+
+    /// Borrow all nodes that are needed to render node node_id.
+    pub fn node_render_pass<'a>(
+        &'a mut self,
+        node_id: &NodeId,
+    ) -> Result<NodeRenderPass<'a>, SceneError> {
+        let input_ids: Vec<NodeId> = self.node(node_id)?.inputs.to_vec();
+        let mut node_and_inputs: HashMap<&NodeId, &mut Node> = self
+            .nodes
+            .iter_mut()
+            .filter(|(id, _node)| input_ids.contains(id) || *id == node_id)
+            .collect();
+        let node = node_and_inputs
+            .remove(&node_id)
+            .ok_or_else(|| SceneError::MissingNode(node_id.clone()))?;
+        let inputs: Vec<&Node> = input_ids
+            .into_iter()
+            .map(|input_id| {
+                node_and_inputs
+                    .remove(&input_id)
+                    .ok_or_else(|| SceneError::MissingNode(node_id.clone()))
+                    .map(|r| &*r)
```

Comment:
not necessary, but I wanted to explicitly cast down &mut Node to &Node for inputs